### PR TITLE
[FLINK-17218][tests] Adding recoverable failures and correctness chec…

### PR DESCRIPTION
**Pull request by AHeise**
…ks to UnalignedCheckpointITCase.

## What is the purpose of the change

This pull request adds recovery logic and verification to integration tests of unaligned checkpoints.

## Brief change log

- Expand the UnalignedCheckpointITCase with induced failures, recovery, and correctness checks.

## Verifying this change

Only changes in tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (not applicable / docs / **JavaDocs** / not documented)

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/api/serialization/SpillingAdaptiveSpanningRecordDeserializer.java [hunk 1]
```diff
@@ -597,12 +597,12 @@ private void addNextChunkFromMemorySegment(MemorySegment segment, int offset, in
 				throw new UnsupportedOperationException("Unaligned checkpoint currently do not support spilled " +
 					"records.");
 			} else if (recordLength != -1) {
-				int leftOverSize = leftOverLimit - leftOverStart;
+				int leftOverSize = leftOverData != null ? leftOverLimit - leftOverStart : 0;
```
Comment (zhijiangW) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/api/serialization/SpillingAdaptiveSpanningRecordDeserializer.java:600]: Thanks for finding this bug!

I think the root cause was the state inconsistency among `{leftOverLimit, leftOverStart}` with `leftOverData`. During `#clear()` we only reset `leftOverData` as null, but now reset the derived `{leftOverLimit, leftOverStart}` from `leftOverData`. So we can check the condition only by `leftOverData`. Maybe we can also reset `{leftOverLimit, leftOverStart}` during `#clear()` to keep all consistency.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SubtaskCheckpointCoordinatorImpl.java [hunk 2]
```diff
@@ -240,7 +240,8 @@ private void finishAndReportAsync(Map<OperatorID, OperatorSnapshotFutures> snaps
 		final Future<?> channelWrittenFuture;
 		if (unalignedCheckpointEnabled) {
 			ChannelStateWriteResult writeResult = channelStateWriter.getWriteResult(metadata.getCheckpointId());
-			channelWrittenFuture = writeResult.getJointFuture();
+			channelWrittenFuture = writeResult.getJointFuture()
+				.whenComplete((dummy, ex) -> channelStateWriter.notifyCheckpointComplete(metadata.getCheckpointId()));
```
[the remark this answers is not in the document]
Comment (AHeise) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SubtaskCheckpointCoordinatorImpl.java]: I called it `stop` now (symmetric to `start`).

**Comment by AHeise** [comment]
> LGTM - I guess commits should be either squashed or renamed from `hotfix` to `[FLINK-17218](https://issues.apache.org/jira/browse/FLINK-17218)`?

Renamed to FLINK-17218.

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriterImpl.java [hunk 3]
```diff
@@ -91,6 +91,7 @@ public ChannelStateWriterImpl(CheckpointStorageWorkerView streamFactoryResolver)
 
 	@Override
 	public void start(long checkpointId, CheckpointOptions checkpointOptions) {
+		results.keySet().forEach(oldCheckpointId -> abort(oldCheckpointId, new Exception("Starting new checkpoint " + checkpointId)));
```
Comment (zhijiangW) [inline flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriterImpl.java:94]: I guess the below `Preconditions.checkState(results.size() < maxCheckpoints, "results.size() > maxCheckpoints", results.size(), maxCheckpoints);` might be invalid if we remove the pending ones here?

Comment (AHeise) [reply flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriterImpl.java:94]: You are right that it's probably not needed. However, it's an inexpensive check that verifies the highlighted line.
I'd keep the check for now.

**Comment by AHeise** [comment]
@flinkbot run azure

**Pull request merged**